### ISSUE TITLE
Scripts of Answer Cleansing

### ISSUE THREAD

**Issue by wang-yiwei** [itsnamgyu/reasoning-teacher#2]
Firstly, thanks for your great work and the open-sourced code! This is definitely a work that deserves to be accepted by the ACL :))

My question about the paper is, how do you semantically chunk the results generated by the student model?
If the sequence generated by the student model ends up including a long string containing rationale and prediction (assuming I'm doing a seq2seq task like text paraphrasing), the small model may not strictly follow the prompt template of "so, the answer is. ", so how do you convert the prediction into the form needed for the evaluation?

I have looked at the `A.3 Answer Cleansing`, and it's said that you followed ”Kojima et al. (2022) to cleanse answers generated by models to assess their correctness“

I am very interested in this cleansing processing, would you mind please show me the script for this step? Or if it's not in the repo, would you mind offering me some insight into how I can convert the answers generated by models into an evaluable format?

Thank you very much && 감사합니다 :))

![image](https://github.com/itsnamgyu/reasoning-teacher/assets/50556116/ebc4f1d0-0bb9-4140-9f71-d4a7787956ff)

**Comment by itsnamgyu** [comment]
Thanks for your interest :)

By default, we fine-tune student models to output CoT predictions in the form of `<REASONING> --> <ANSWER>`.

- If the student model follows the correct format, we assume that the answer comes right after the prediction prefix (`-->`) and take the _first_ answer candidate. E.g., if it is a math dataset, we take the first number that comes after `-->`.
- If the student model doesn't output the correct format, i.e., doesn't output `-->`, then we assume that the answer comes at the end of `<ANSWER>` and take the _last_ answer candidate.

The code for this is in `evaluation.evaluator.Evaluator.cleanse_prediction` (in the `src/` folder). The example above assumes `Evaluator.prediction_template == "ft_cot_token"`. Other methods use different prediction templates, with different prediction prefixes. For example, the prediction prefix used for Few-shot-CoT is `The answer is`.

**Comment by wang-yiwei** [comment]
Thank you very much for the answer. I will try to generate labels and CoTs from gpt-3.5 API with the output format requirement `<REASONING> --> <ANSWER>` (especially with the `-->`). And I assume your whole prompts format guiding LLMs to generate CoT-datasets might be something like: `<INSTRUCTION><INPUT><REASONING> --> <ANSWER>`

Regarding the answer cleansing, I appreciate your suggestions pretty much, but I think I might need to adjust the rule-based algorithm firstly, since the data I will deal with is sequence-to-sequence (not a single token or numeric data), therefore it might be even harder to separate the target sequence from a long output string. But the `-->` is definitely a good idea, I will check if it will work in the small models' reasoning results, do hope so though.

Best regards

(P.S. If you don't mind I can keep you updated later on and no worries about the issue, it can stay as closed even after your feedbacks if you don't want to make it open all the time :) thank you very much

**Comment by itsnamgyu** [comment]
Sure feel free to reach out. I do think there is some misunderstanding regarding the methodology, so I would recommend taking a closer look at our methodology section.

- We use Zero-shot-CoT from Kojima et al 2022 for teacher generation (i.e., two-step generation using "let's think step by step" and "therefore the answer is")
- We parse out and format the generated reasoning into the format you mentioned (with the `-->`) and fine-tune the students models with those formatted samples

**Comment by wang-yiwei** [comment]
Hey :)) Namgyu,

I finished the first round of modelling and evaluating my text detoxifier.
However, knowledge distillation with reasoning isn't really ideal if I use the student model to generate chain-of-thoughts and final nontoxic sentences. Especially when I consider the results from the paper [Distilling Step-by-Step](https://twitter.com/cydhsieh/status/1654328897805385728)

![image](https://github.com/itsnamgyu/reasoning-teacher/assets/50556116/a61d7efe-8196-404d-9005-8473fa28741b)

As you can see from the picture, `distill-output-cot` is the model that is trained with prompt template as follows:

![image](https://github.com/itsnamgyu/reasoning-teacher/assets/50556116/e3f5f998-62c6-4dff-b5d9-c08903a5531e)

and the labels:

![image](https://github.com/itsnamgyu/reasoning-teacher/assets/50556116/76d4c9d8-6e31-45e0-9ce8-4f181ea9ffef)

I don't have any output format issue since the language model I used, flan-t5, follows instructions very well.
 
I actually saw your tweet with the author of "Distilling Step-by-Step" paper, and I am also interested in the prompt engineering part of the study, which I think isn't well formulated. So, I am wondering if you would like to have a discussion with me in terms of the prompt engineering in your paper as well as in the "Distilling Step-by-Step" paper sometime when both of us are free?

btw, due to privacy concerns, you can find me via yiwei.wang@uni-konstanz.de

**Comment by itsnamgyu** [comment]
Hi Iwan, I've found that small GPT-3 models <100B are unable to generate good CoT rationales for complex questions, esp. math, but they do form comprehensive answers for text-based questions.

We didn't perform any prompt engineering for our paper. We used the same prompt from the zero-shot cot paper [Kojima 2022] to highlight the versatility of our method. If you would still like to talk about prompt engineering, feel free to send me an email next week! I'm on vacation until next Tuesday 🌝

I didn't understand your table, but smaller students may struggle to learn from long-form reasoning such as the one you've shown. The Distilling Step-by-Step paper uses a multi-task training formulation that trains the model to either (1) answer the question without reasoning or (2) generate step-by-step reasoning, which I think you could consider to boost performance–in theory, it should always outperform standard fine-tuning baselines. You might also want to see SCOTT (https://arxiv.org/abs/2305.01879) (ACL 2023 Oral) which uses multi-task learning to learn from incorrect rationales.
